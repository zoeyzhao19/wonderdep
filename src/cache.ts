// https://github.com/antfu/taze/src/io/resolves.ts
import { resolve } from 'node:path'
import os from 'node:os'
import { existsSync, promises as fs } from 'node:fs'
import { consola } from 'consola'

let cache: Record<string, Record<string, any>> = {}

const cacheDir = resolve(os.tmpdir(), 'wonder-dep')
const cachePath = resolve(cacheDir, 'cache.json')

export async function loadCache() {
  if (existsSync(cachePath)) {
    consola.info(`cache loaded from ${cachePath}`)
    cache = JSON.parse(await fs.readFile(cachePath, 'utf-8'))
  }
  return cache
}

export async function dumpCache() {
  try {
    await fs.mkdir(cacheDir, { recursive: true })
    await fs.writeFile(cachePath, JSON.stringify(cache), 'utf-8')
    consola.info(`cache saved to ${cachePath}`)
  }
  catch (err) {
    consola.warn('Failed to save cache')
    consola.warn(err)
  }
}
