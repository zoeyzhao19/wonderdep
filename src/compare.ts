import { Range, satisfies } from 'semver'
import type { ResolvedVersion } from './resolve'

/**
 * check if target version  satisfy source version restriction
 */
export function compareVersion(target: ResolvedVersion, source: string) {
  try {
    const sourceRange = new Range(source)
    let targetStr = ''
    if (target[4])
      targetStr = `.${target[4]}${targetStr}`

    if (target[3])
      targetStr = `.${target[3]}${targetStr}`

    if (target[2])
      targetStr = `.${target[2]}${targetStr}`

    return satisfies(targetStr.slice(1), sourceRange)
  }
  catch (err: any) {
    throw new Error(err)
  }
}
