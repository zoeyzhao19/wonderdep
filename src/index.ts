export { resolvePkg, resolveDepsVersion } from './resolve'
