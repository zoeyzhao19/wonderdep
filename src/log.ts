import { consola } from 'consola'

const log = consola.withDefaults({
  tag: 'wonderdep',
})

export { log }
