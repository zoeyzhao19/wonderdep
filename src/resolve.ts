import { consola } from 'consola'
import pacote from 'pacote'
import { valid } from 'semver'
import { compareVersion } from './compare'
import { dumpCache, loadCache } from './cache'

export type ResolvedVersion = [string, '^' | '~' | '=' | '<' | '>' | '<=' | '>=', string | undefined, string | undefined, string | undefined]

function ensureSpecifyDep(dep: string) {
  return dep.replace(/(\S+?)[\^~@](.*)/, '$1=$2')
}

async function resolveHostPkgPack(name: string) {
  consola.info(`üç≥ resolving all available versions of ${name}...`)
  const pack = await pacote.packument(name)
  return pack
}

export async function resolvePkgManifest(name: string) {
  consola.info(`üç≥ resolving manifest for ${name}...`)
  const manifest = await pacote.manifest(name)
  return manifest
}

function resolveVersion(pkg: string): ResolvedVersion | undefined {
  const [pkgName, pkgVersion] = pkg.split(/(?:<=)|(?:>=)|(?:[\^~=<>])/)

  let type: any = pkg.slice(pkgName.length, pkgName.length + 1)

  if (!valid(pkgVersion))
    return

  if (!type || type === '@')
    type = '='

  else if (!['=', '^', '~', '<=', '>='].includes(type))
    consola.error(`The type ${type} of ${pkgName} is not supported`)

  if (!pkgVersion)
    return [pkgName, type, undefined, undefined, undefined]

  const [release] = pkgVersion.split(/[-+]/)
  const [major, minor, patch] = release.split('.')

  return [pkgName, type, major, minor, patch]
}

/**
 *
 * @deprecated use `resolveDepsVersion` instead
 */
export async function resolvePkg(hostPkgNameWithV: string, deps: string[]) {
  return resolveHostVersion(hostPkgNameWithV, deps)
}

export async function resolveHostVersion(hostPkgName: string, deps: string[]) {
  const pack = await resolveHostPkgPack(hostPkgName)

  const cache = await loadCache()
  let shouldUpdateCache = false

  const result: string[] = []
  for (let dep of deps) {
    dep = ensureSpecifyDep(dep)

    const resolved = resolveVersion(dep) // resolve dep with version info [pkgName, type, major, minor, patch]
    if (!resolved) {
      consola.info(`The dep ${dep} is not valid, skip`)
      break
    }

    const packageVersions = Object.keys(pack.versions).reverse()
    let found = false
    for (const packageVersion of packageVersions) {
      if (found)
        break

      let manifestDeps: Record<string, string>
      const packageWithVersion = `${hostPkgName}@${packageVersion}`
      if (!cache[`${packageWithVersion}`]) {
        const hostManifest = await resolvePkgManifest(`${packageWithVersion}`)
        manifestDeps = { ...hostManifest.dependencies, ...hostManifest.devDependencies, ...hostManifest.optionalDependencies }
        shouldUpdateCache = true
        cache[`${packageWithVersion}`] = {
          name: hostManifest.name,
          version: hostManifest.version,
          dependencies: hostManifest.dependencies,
          devDependencies: hostManifest.devDependencies,
          optionalDependencies: hostManifest.optionalDependencies,
        }
      }
      else {
        const cacheInfo = cache[`${packageWithVersion}`]
        manifestDeps = { ...cacheInfo.dependencies, ...cacheInfo.devDependencies, ...cacheInfo.optionalDependencies }
      }

      const cacheName = cache[`${packageWithVersion}`].name
      const cacheVersion = cache[`${packageWithVersion}`].version

      if (manifestDeps[resolved[0]]) {
        if (compareVersion(resolved, manifestDeps[resolved[0]])) {
          result.push(`${cacheName}=${cacheVersion} => ${resolved[0]}${manifestDeps[resolved[0]]?.trim()}`)
          found = true
          break
        }
      }
      else {
        consola.info(`The dep ${dep} is not specified in ${packageWithVersion}, abort to resolve more`)
        break
      }
    }
  }

  shouldUpdateCache && await dumpCache()

  if (result.length) {
    consola.success('package resolved succeed')
    consola.success(`Found the nearest ${hostPkgName} version for ${deps.join(', ')}`)
    consola.success(`${result.join('\n  ')}\n`)
  }

  return result
}
