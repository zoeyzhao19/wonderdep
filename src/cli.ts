import cac from 'cac'
import { consola } from 'consola'
import { resolveHostVersion } from './resolve'

const cli = cac('wonder')

cli.command('<packageName>', 'raw package name, and should not append with a version. e.g: vitest')
  .option('--deps <deps>', 'deps which is depended, should append with a semver version')
  .action((packageName, options) => {
    if (!options.deps) {
      consola.warn('No deps provided')
      return
    }
    const deps = options.deps || ''
    resolveHostVersion(packageName, deps.split(' '))
  })

cli.help()
cli.parse()
