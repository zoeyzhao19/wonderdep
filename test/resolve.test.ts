import { describe, expect, test } from 'vitest'
import { resolveHostVersion } from '../src'

describe('resolve', async () => {
  test('resolve specify host package version from a specify dep', async () => {
    const data: [string, string[]] = ['vite-plugin-inspect', ['vite@4.4.8']]

    const result = await resolveHostVersion(data[0], data[1])
    expect(result).toMatchInlineSnapshot(`
      [
        "vite-plugin-inspect=0.7.35 => vite^4.4.8",
      ]
    `)
  }, {
    timeout: 3000,
  })
})
