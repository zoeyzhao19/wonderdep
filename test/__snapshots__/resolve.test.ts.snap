// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`resolve > resolveHostPkgManifest 1`] = `
[
  {
    "_from": "vite@4.4.8",
    "_id": "vite@4.4.8",
    "_integrity": "sha512-LONawOUUjxQridNWGQlNizfKH89qPigK36XhMI7COMGztz8KNY0JHim7/xDd71CZwGT4HtSRgI7Hy+RlhG0Gvg==",
    "_resolved": "https://registry.npmjs.org/vite/-/vite-4.4.8.tgz",
    "_signatures": [
      {
        "keyid": "SHA256:jl3bwswu80PjjokCgh0o2w5c2U4LhQAE57gj9cz1kzA",
        "sig": "MEQCIA7Lre3hTEInyzhooq68Mk1GOQi9k5DavTHQczY6CkSQAiBxEc79LBHAdwr29/9O6ksZCLJaxkDUKraV/Es+NopfOw==",
      },
    ],
    "bin": {
      "vite": "bin/vite.js",
    },
    "dependencies": {
      "esbuild": "^0.18.10",
      "postcss": "^8.4.26",
      "rollup": "^3.25.2",
    },
    "devDependencies": {
      "@ampproject/remapping": "^2.2.1",
      "@babel/parser": "^7.22.7",
      "@babel/types": "^7.22.5",
      "@jridgewell/trace-mapping": "^0.3.18",
      "@rollup/plugin-alias": "^4.0.4",
      "@rollup/plugin-commonjs": "^25.0.3",
      "@rollup/plugin-dynamic-import-vars": "^2.0.4",
      "@rollup/plugin-json": "^6.0.0",
      "@rollup/plugin-node-resolve": "15.1.0",
      "@rollup/plugin-typescript": "^11.1.2",
      "@rollup/pluginutils": "^5.0.2",
      "@types/escape-html": "^1.0.2",
      "@types/pnpapi": "^0.0.2",
      "acorn": "^8.10.0",
      "acorn-walk": "^8.2.0",
      "cac": "^6.7.14",
      "chokidar": "^3.5.3",
      "connect": "^3.7.0",
      "connect-history-api-fallback": "^2.0.0",
      "convert-source-map": "^2.0.0",
      "cors": "^2.8.5",
      "cross-spawn": "^7.0.3",
      "debug": "^4.3.4",
      "dep-types": "link:./src/types",
      "dotenv": "^16.3.1",
      "dotenv-expand": "^9.0.0",
      "es-module-lexer": "^1.3.0",
      "escape-html": "^1.0.3",
      "estree-walker": "^3.0.3",
      "etag": "^1.8.1",
      "fast-glob": "^3.3.0",
      "http-proxy": "^1.18.1",
      "json-stable-stringify": "^1.0.2",
      "launch-editor-middleware": "^2.6.0",
      "lightningcss": "^1.21.5",
      "magic-string": "^0.30.2",
      "micromatch": "^4.0.5",
      "mlly": "^1.4.0",
      "mrmime": "^1.0.1",
      "okie": "^1.0.1",
      "open": "^8.4.2",
      "parse5": "^7.1.2",
      "periscopic": "^3.1.0",
      "picocolors": "^1.0.0",
      "picomatch": "^2.3.1",
      "postcss-import": "^15.1.0",
      "postcss-load-config": "^4.0.1",
      "postcss-modules": "^6.0.0",
      "resolve.exports": "^2.0.2",
      "rollup-plugin-license": "^3.0.1",
      "sirv": "^2.0.3",
      "source-map-support": "^0.5.21",
      "strip-ansi": "^7.1.0",
      "strip-literal": "^1.0.1",
      "tsconfck": "^2.1.2",
      "tslib": "^2.6.0",
      "types": "link:./types",
      "ufo": "^1.1.2",
      "ws": "^8.13.0",
    },
    "dist": {
      "fileCount": 28,
      "integrity": "sha512-LONawOUUjxQridNWGQlNizfKH89qPigK36XhMI7COMGztz8KNY0JHim7/xDd71CZwGT4HtSRgI7Hy+RlhG0Gvg==",
      "shasum": "31e4a438f8748695c68bd57ffd262ba93540fdf7",
      "signatures": [
        {
          "keyid": "SHA256:jl3bwswu80PjjokCgh0o2w5c2U4LhQAE57gj9cz1kzA",
          "sig": "MEQCIA7Lre3hTEInyzhooq68Mk1GOQi9k5DavTHQczY6CkSQAiBxEc79LBHAdwr29/9O6ksZCLJaxkDUKraV/Es+NopfOw==",
        },
      ],
      "tarball": "https://registry.npmjs.org/vite/-/vite-4.4.8.tgz",
      "unpackedSize": 3333060,
    },
    "engines": {
      "node": "^14.18.0 || >=16.0.0",
    },
    "funding": {
      "url": "https://github.com/vitejs/vite?sponsor=1",
    },
    "name": "vite",
    "optionalDependencies": {
      "fsevents": "~2.3.2",
    },
    "peerDependencies": {
      "@types/node": ">= 14",
      "less": "*",
      "lightningcss": "^1.21.0",
      "sass": "*",
      "stylus": "*",
      "sugarss": "*",
      "terser": "^5.4.0",
    },
    "peerDependenciesMeta": {
      "@types/node": {
        "optional": true,
      },
      "less": {
        "optional": true,
      },
      "lightningcss": {
        "optional": true,
      },
      "sass": {
        "optional": true,
      },
      "stylus": {
        "optional": true,
      },
      "sugarss": {
        "optional": true,
      },
      "terser": {
        "optional": true,
      },
    },
    "version": "4.4.8",
  },
  {
    "_attestations": {
      "provenance": {
        "predicateType": "https://slsa.dev/provenance/v1",
      },
      "url": "https://registry.npmjs.org/-/npm/v1/attestations/vite@5.0.10",
    },
    "_from": "vite@5",
    "_id": "vite@5.0.10",
    "_integrity": "sha512-2P8J7WWgmc355HUMlFrwofacvr98DAjoE52BfdbwQtyLH06XKwaL/FMnmKM2crF0iX4MpmMKoDlNCB1ok7zHCw==",
    "_resolved": "https://registry.npmjs.org/vite/-/vite-5.0.10.tgz",
    "_signatures": [
      {
        "keyid": "SHA256:jl3bwswu80PjjokCgh0o2w5c2U4LhQAE57gj9cz1kzA",
        "sig": "MEYCIQDiqbTtPXFeKFVUkGmpMvQdc9z3mP3ZKC8g9dgm/9ug7gIhANMS8KaG1OoLv7x/4Hg7W1v26yEUTF8A374UJ41VHnb/",
      },
    ],
    "bin": {
      "vite": "bin/vite.js",
    },
    "dependencies": {
      "esbuild": "^0.19.3",
      "postcss": "^8.4.32",
      "rollup": "^4.2.0",
    },
    "devDependencies": {
      "@ampproject/remapping": "^2.2.1",
      "@babel/parser": "^7.23.6",
      "@jridgewell/trace-mapping": "^0.3.20",
      "@rollup/plugin-alias": "^5.1.0",
      "@rollup/plugin-commonjs": "^25.0.7",
      "@rollup/plugin-dynamic-import-vars": "^2.1.2",
      "@rollup/plugin-json": "^6.0.1",
      "@rollup/plugin-node-resolve": "15.2.3",
      "@rollup/plugin-typescript": "^11.1.5",
      "@rollup/pluginutils": "^5.1.0",
      "@types/escape-html": "^1.0.4",
      "@types/pnpapi": "^0.0.5",
      "acorn": "^8.11.2",
      "acorn-walk": "^8.3.1",
      "cac": "^6.7.14",
      "chokidar": "^3.5.3",
      "connect": "^3.7.0",
      "convert-source-map": "^2.0.0",
      "cors": "^2.8.5",
      "cross-spawn": "^7.0.3",
      "debug": "^4.3.4",
      "dep-types": "link:./src/types",
      "dotenv": "^16.3.1",
      "dotenv-expand": "^10.0.0",
      "es-module-lexer": "^1.4.1",
      "escape-html": "^1.0.3",
      "estree-walker": "^3.0.3",
      "etag": "^1.8.1",
      "fast-glob": "^3.3.2",
      "http-proxy": "^1.18.1",
      "json-stable-stringify": "^1.1.0",
      "launch-editor-middleware": "^2.6.1",
      "lightningcss": "^1.22.1",
      "magic-string": "^0.30.5",
      "micromatch": "^4.0.5",
      "mlly": "^1.4.2",
      "mrmime": "^1.0.1",
      "okie": "^1.0.1",
      "open": "^8.4.2",
      "parse5": "^7.1.2",
      "periscopic": "^4.0.2",
      "picocolors": "^1.0.0",
      "picomatch": "^2.3.1",
      "postcss-import": "^15.1.0",
      "postcss-load-config": "^4.0.2",
      "postcss-modules": "^6.0.0",
      "resolve.exports": "^2.0.2",
      "rollup-plugin-dts": "^6.1.0",
      "rollup-plugin-license": "^3.2.0",
      "sirv": "^2.0.3",
      "source-map-support": "^0.5.21",
      "strip-ansi": "^7.1.0",
      "strip-literal": "^1.3.0",
      "tsconfck": "^3.0.0",
      "tslib": "^2.6.2",
      "types": "link:./types",
      "ufo": "^1.3.2",
      "ws": "^8.15.0",
    },
    "dist": {
      "attestations": {
        "provenance": {
          "predicateType": "https://slsa.dev/provenance/v1",
        },
        "url": "https://registry.npmjs.org/-/npm/v1/attestations/vite@5.0.10",
      },
      "fileCount": 30,
      "integrity": "sha512-2P8J7WWgmc355HUMlFrwofacvr98DAjoE52BfdbwQtyLH06XKwaL/FMnmKM2crF0iX4MpmMKoDlNCB1ok7zHCw==",
      "shasum": "1e13ef5c3cf5aa4eed81f5df6d107b3c3f1f6356",
      "signatures": [
        {
          "keyid": "SHA256:jl3bwswu80PjjokCgh0o2w5c2U4LhQAE57gj9cz1kzA",
          "sig": "MEYCIQDiqbTtPXFeKFVUkGmpMvQdc9z3mP3ZKC8g9dgm/9ug7gIhANMS8KaG1OoLv7x/4Hg7W1v26yEUTF8A374UJ41VHnb/",
        },
      ],
      "tarball": "https://registry.npmjs.org/vite/-/vite-5.0.10.tgz",
      "unpackedSize": 3473932,
    },
    "engines": {
      "node": "^18.0.0 || >=20.0.0",
    },
    "funding": {
      "url": "https://github.com/vitejs/vite?sponsor=1",
    },
    "name": "vite",
    "optionalDependencies": {
      "fsevents": "~2.3.3",
    },
    "peerDependencies": {
      "@types/node": "^18.0.0 || >=20.0.0",
      "less": "*",
      "lightningcss": "^1.21.0",
      "sass": "*",
      "stylus": "*",
      "sugarss": "*",
      "terser": "^5.4.0",
    },
    "peerDependenciesMeta": {
      "@types/node": {
        "optional": true,
      },
      "less": {
        "optional": true,
      },
      "lightningcss": {
        "optional": true,
      },
      "sass": {
        "optional": true,
      },
      "stylus": {
        "optional": true,
      },
      "sugarss": {
        "optional": true,
      },
      "terser": {
        "optional": true,
      },
    },
    "version": "5.0.10",
  },
]
`;

exports[`resolve > resolvePkgManifest 1`] = `
[
  {
    "_from": "vite@4.4.8",
    "_id": "vite@4.4.8",
    "_integrity": "sha512-LONawOUUjxQridNWGQlNizfKH89qPigK36XhMI7COMGztz8KNY0JHim7/xDd71CZwGT4HtSRgI7Hy+RlhG0Gvg==",
    "_resolved": "https://registry.npmjs.org/vite/-/vite-4.4.8.tgz",
    "_signatures": [
      {
        "keyid": "SHA256:jl3bwswu80PjjokCgh0o2w5c2U4LhQAE57gj9cz1kzA",
        "sig": "MEQCIA7Lre3hTEInyzhooq68Mk1GOQi9k5DavTHQczY6CkSQAiBxEc79LBHAdwr29/9O6ksZCLJaxkDUKraV/Es+NopfOw==",
      },
    ],
    "bin": {
      "vite": "bin/vite.js",
    },
    "dependencies": {
      "esbuild": "^0.18.10",
      "postcss": "^8.4.26",
      "rollup": "^3.25.2",
    },
    "devDependencies": {
      "@ampproject/remapping": "^2.2.1",
      "@babel/parser": "^7.22.7",
      "@babel/types": "^7.22.5",
      "@jridgewell/trace-mapping": "^0.3.18",
      "@rollup/plugin-alias": "^4.0.4",
      "@rollup/plugin-commonjs": "^25.0.3",
      "@rollup/plugin-dynamic-import-vars": "^2.0.4",
      "@rollup/plugin-json": "^6.0.0",
      "@rollup/plugin-node-resolve": "15.1.0",
      "@rollup/plugin-typescript": "^11.1.2",
      "@rollup/pluginutils": "^5.0.2",
      "@types/escape-html": "^1.0.2",
      "@types/pnpapi": "^0.0.2",
      "acorn": "^8.10.0",
      "acorn-walk": "^8.2.0",
      "cac": "^6.7.14",
      "chokidar": "^3.5.3",
      "connect": "^3.7.0",
      "connect-history-api-fallback": "^2.0.0",
      "convert-source-map": "^2.0.0",
      "cors": "^2.8.5",
      "cross-spawn": "^7.0.3",
      "debug": "^4.3.4",
      "dep-types": "link:./src/types",
      "dotenv": "^16.3.1",
      "dotenv-expand": "^9.0.0",
      "es-module-lexer": "^1.3.0",
      "escape-html": "^1.0.3",
      "estree-walker": "^3.0.3",
      "etag": "^1.8.1",
      "fast-glob": "^3.3.0",
      "http-proxy": "^1.18.1",
      "json-stable-stringify": "^1.0.2",
      "launch-editor-middleware": "^2.6.0",
      "lightningcss": "^1.21.5",
      "magic-string": "^0.30.2",
      "micromatch": "^4.0.5",
      "mlly": "^1.4.0",
      "mrmime": "^1.0.1",
      "okie": "^1.0.1",
      "open": "^8.4.2",
      "parse5": "^7.1.2",
      "periscopic": "^3.1.0",
      "picocolors": "^1.0.0",
      "picomatch": "^2.3.1",
      "postcss-import": "^15.1.0",
      "postcss-load-config": "^4.0.1",
      "postcss-modules": "^6.0.0",
      "resolve.exports": "^2.0.2",
      "rollup-plugin-license": "^3.0.1",
      "sirv": "^2.0.3",
      "source-map-support": "^0.5.21",
      "strip-ansi": "^7.1.0",
      "strip-literal": "^1.0.1",
      "tsconfck": "^2.1.2",
      "tslib": "^2.6.0",
      "types": "link:./types",
      "ufo": "^1.1.2",
      "ws": "^8.13.0",
    },
    "dist": {
      "fileCount": 28,
      "integrity": "sha512-LONawOUUjxQridNWGQlNizfKH89qPigK36XhMI7COMGztz8KNY0JHim7/xDd71CZwGT4HtSRgI7Hy+RlhG0Gvg==",
      "shasum": "31e4a438f8748695c68bd57ffd262ba93540fdf7",
      "signatures": [
        {
          "keyid": "SHA256:jl3bwswu80PjjokCgh0o2w5c2U4LhQAE57gj9cz1kzA",
          "sig": "MEQCIA7Lre3hTEInyzhooq68Mk1GOQi9k5DavTHQczY6CkSQAiBxEc79LBHAdwr29/9O6ksZCLJaxkDUKraV/Es+NopfOw==",
        },
      ],
      "tarball": "https://registry.npmjs.org/vite/-/vite-4.4.8.tgz",
      "unpackedSize": 3333060,
    },
    "engines": {
      "node": "^14.18.0 || >=16.0.0",
    },
    "funding": {
      "url": "https://github.com/vitejs/vite?sponsor=1",
    },
    "name": "vite",
    "optionalDependencies": {
      "fsevents": "~2.3.2",
    },
    "peerDependencies": {
      "@types/node": ">= 14",
      "less": "*",
      "lightningcss": "^1.21.0",
      "sass": "*",
      "stylus": "*",
      "sugarss": "*",
      "terser": "^5.4.0",
    },
    "peerDependenciesMeta": {
      "@types/node": {
        "optional": true,
      },
      "less": {
        "optional": true,
      },
      "lightningcss": {
        "optional": true,
      },
      "sass": {
        "optional": true,
      },
      "stylus": {
        "optional": true,
      },
      "sugarss": {
        "optional": true,
      },
      "terser": {
        "optional": true,
      },
    },
    "version": "4.4.8",
  },
  {
    "_attestations": {
      "provenance": {
        "predicateType": "https://slsa.dev/provenance/v1",
      },
      "url": "https://registry.npmjs.org/-/npm/v1/attestations/vite@5.0.10",
    },
    "_from": "vite@5",
    "_id": "vite@5.0.10",
    "_integrity": "sha512-2P8J7WWgmc355HUMlFrwofacvr98DAjoE52BfdbwQtyLH06XKwaL/FMnmKM2crF0iX4MpmMKoDlNCB1ok7zHCw==",
    "_resolved": "https://registry.npmjs.org/vite/-/vite-5.0.10.tgz",
    "_signatures": [
      {
        "keyid": "SHA256:jl3bwswu80PjjokCgh0o2w5c2U4LhQAE57gj9cz1kzA",
        "sig": "MEYCIQDiqbTtPXFeKFVUkGmpMvQdc9z3mP3ZKC8g9dgm/9ug7gIhANMS8KaG1OoLv7x/4Hg7W1v26yEUTF8A374UJ41VHnb/",
      },
    ],
    "bin": {
      "vite": "bin/vite.js",
    },
    "dependencies": {
      "esbuild": "^0.19.3",
      "postcss": "^8.4.32",
      "rollup": "^4.2.0",
    },
    "devDependencies": {
      "@ampproject/remapping": "^2.2.1",
      "@babel/parser": "^7.23.6",
      "@jridgewell/trace-mapping": "^0.3.20",
      "@rollup/plugin-alias": "^5.1.0",
      "@rollup/plugin-commonjs": "^25.0.7",
      "@rollup/plugin-dynamic-import-vars": "^2.1.2",
      "@rollup/plugin-json": "^6.0.1",
      "@rollup/plugin-node-resolve": "15.2.3",
      "@rollup/plugin-typescript": "^11.1.5",
      "@rollup/pluginutils": "^5.1.0",
      "@types/escape-html": "^1.0.4",
      "@types/pnpapi": "^0.0.5",
      "acorn": "^8.11.2",
      "acorn-walk": "^8.3.1",
      "cac": "^6.7.14",
      "chokidar": "^3.5.3",
      "connect": "^3.7.0",
      "convert-source-map": "^2.0.0",
      "cors": "^2.8.5",
      "cross-spawn": "^7.0.3",
      "debug": "^4.3.4",
      "dep-types": "link:./src/types",
      "dotenv": "^16.3.1",
      "dotenv-expand": "^10.0.0",
      "es-module-lexer": "^1.4.1",
      "escape-html": "^1.0.3",
      "estree-walker": "^3.0.3",
      "etag": "^1.8.1",
      "fast-glob": "^3.3.2",
      "http-proxy": "^1.18.1",
      "json-stable-stringify": "^1.1.0",
      "launch-editor-middleware": "^2.6.1",
      "lightningcss": "^1.22.1",
      "magic-string": "^0.30.5",
      "micromatch": "^4.0.5",
      "mlly": "^1.4.2",
      "mrmime": "^1.0.1",
      "okie": "^1.0.1",
      "open": "^8.4.2",
      "parse5": "^7.1.2",
      "periscopic": "^4.0.2",
      "picocolors": "^1.0.0",
      "picomatch": "^2.3.1",
      "postcss-import": "^15.1.0",
      "postcss-load-config": "^4.0.2",
      "postcss-modules": "^6.0.0",
      "resolve.exports": "^2.0.2",
      "rollup-plugin-dts": "^6.1.0",
      "rollup-plugin-license": "^3.2.0",
      "sirv": "^2.0.3",
      "source-map-support": "^0.5.21",
      "strip-ansi": "^7.1.0",
      "strip-literal": "^1.3.0",
      "tsconfck": "^3.0.0",
      "tslib": "^2.6.2",
      "types": "link:./types",
      "ufo": "^1.3.2",
      "ws": "^8.15.0",
    },
    "dist": {
      "attestations": {
        "provenance": {
          "predicateType": "https://slsa.dev/provenance/v1",
        },
        "url": "https://registry.npmjs.org/-/npm/v1/attestations/vite@5.0.10",
      },
      "fileCount": 30,
      "integrity": "sha512-2P8J7WWgmc355HUMlFrwofacvr98DAjoE52BfdbwQtyLH06XKwaL/FMnmKM2crF0iX4MpmMKoDlNCB1ok7zHCw==",
      "shasum": "1e13ef5c3cf5aa4eed81f5df6d107b3c3f1f6356",
      "signatures": [
        {
          "keyid": "SHA256:jl3bwswu80PjjokCgh0o2w5c2U4LhQAE57gj9cz1kzA",
          "sig": "MEYCIQDiqbTtPXFeKFVUkGmpMvQdc9z3mP3ZKC8g9dgm/9ug7gIhANMS8KaG1OoLv7x/4Hg7W1v26yEUTF8A374UJ41VHnb/",
        },
      ],
      "tarball": "https://registry.npmjs.org/vite/-/vite-5.0.10.tgz",
      "unpackedSize": 3473932,
    },
    "engines": {
      "node": "^18.0.0 || >=20.0.0",
    },
    "funding": {
      "url": "https://github.com/vitejs/vite?sponsor=1",
    },
    "name": "vite",
    "optionalDependencies": {
      "fsevents": "~2.3.3",
    },
    "peerDependencies": {
      "@types/node": "^18.0.0 || >=20.0.0",
      "less": "*",
      "lightningcss": "^1.21.0",
      "sass": "*",
      "stylus": "*",
      "sugarss": "*",
      "terser": "^5.4.0",
    },
    "peerDependenciesMeta": {
      "@types/node": {
        "optional": true,
      },
      "less": {
        "optional": true,
      },
      "lightningcss": {
        "optional": true,
      },
      "sass": {
        "optional": true,
      },
      "stylus": {
        "optional": true,
      },
      "sugarss": {
        "optional": true,
      },
      "terser": {
        "optional": true,
      },
    },
    "version": "5.0.10",
  },
]
`;
